quinn-proto/src/connection.rs

```diff
@@ -807,24 +793,34 @@ impl Connection {
             }
             Err(None) => {
                 debug!(self.log, "failed to authenticate packet");
-                return;
+                if self.check_for_stateless_reset(mux, &packet) {
+                    Err(ConnectionError::Reset)
+                } else {
+                    return;
+                }
             }
             Ok(number) => {
-                if let Some(number) = number {
-                    if self.dedup.insert(number) {
+                let duplicate =
+                    number.and_then(|n| if self.dedup.insert(n) { Some(n) } else { None });
+
+                if let Some(number) = duplicate {
+                    if self.check_for_stateless_reset(mux, &packet) {
```

Comment:
Wait, this can't be right. Stateless resets are never encrypted, because the peer cannot possibly have the connection's keys if it's sending one. I guess we need to `check_for_stateless_reset` *before* decrypting, but only inspect its return value afterwards. It's confusing that the draft allows for the possibility that a stateless reset message just coincidentally happens to be decryptable, so maybe I'm missing something here...

Reply:
I'm not sure why you think it's wrong? We try to decrypt it, which makes sense to me, then when that fails without a transport error we check for the stateless reset.

Arguably the case where the decrypt succeeds but then turns out to be a duplicate packet is very unlikely but technically possible (I think), and we handle that as well.

Reply:
To elaborate on my unease here: a genuine duplicate packet can presumably happen due to network hardware glitchiness on the path between two hosts, while a stateless reset packet that just happens to pass strong cryptographic authentication seems like one of those not-in-the-lifetime-of-the-universe things.